src/vsr/superblock_free_set_encoded.zig

```diff
@@ -0,0 +1,431 @@
+            assert(set.block_index < set.block_count);
+            assert((set.size_transferred == 0) == (set.block_index == 0));
+
+            const chunk = Chunk.for_block(.{
+                .block_index = set.block_index,
+                .block_count = set.block_count,
+                .free_set_size = set.size,
+            });
+
+            const header = mem.bytesAsValue(
+                vsr.Header.Block,
+                set.write_block[0..@sizeOf(vsr.Header)],
+            );
+            header.* = .{
+                .cluster = set.grid.?.superblock.working.cluster,
+                .metadata_bytes = @bitCast(if (set.block_index == 0)
+                    schema.FreeSetNode.Metadata{
+                        .next_free_set_block_checksum = 0,
+                        .next_free_set_block_address = 0,
+                    }
+                else
+                    schema.FreeSetNode.Metadata{
+                        .next_free_set_block_checksum = set.block_checksums[set.block_index - 1],
+                        .next_free_set_block_address = set.block_addresses[set.block_index - 1],
+                    }),
+                .address = set.block_addresses[set.block_index],
+                .snapshot = 0, // TODO(snapshots): Set this properly; it is useful for debugging.
+                .size = @sizeOf(vsr.Header) + chunk.size,
+                .command = .block,
+                .block_type = .free_set,
+            };
+            stdx.copy_disjoint(
+                .exact,
+                u8,
+                set.write_block[@sizeOf(vsr.Header)..][0..chunk.size],
+                set.buffer[chunk.start..chunk.end],
+            );
+            set.size_transferred += chunk.size;
+            header.set_checksum_body(set.write_block[@sizeOf(vsr.Header)..][0..chunk.size]);
+            header.set_checksum();
+            schema.FreeSetNode.assert_valid_header(set.write_block);
+
+            set.block_checksums[set.block_index] = header.checksum;
+            set.grid.?.create_block(
+                checkpoint_write_next_callback,
+                &set.write,
+                &set.write_block,
+            );
+        }
+
+        fn checkpoint_write_next_callback(write: *Grid.Write) void {
+            const set = @fieldParentPtr(Self, "write", write);
+            assert(set.callback == .checkpoint);
+
+            if (set.block_index == set.block_count - 1) {
+                set.checkpoint_done();
+            } else {
+                set.block_index += 1;
+                set.checkpoint_write_next();
+            }
+        }
+
+        fn checkpoint_done(set: *Self) void {
+            assert(set.callback == .checkpoint);
+            defer assert(set.callback == .none);
+
+            if (set.block_count == 0) {
+                assert(set.size == 0);
+                assert(set.block_index == 0);
+            } else {
+                assert(set.block_index == set.block_count - 1);
+            }
```

Comment:
I initially avoided that because then the reverse case, in the open, looks weird, as the index is one greater than the real index. But that's actually an idiomatic form of a downwards counted loop, so switched to that to avoid special cases.